File: c++/jan/25_24/add_two_numbers.cpp
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        // Create a dummy node to simplify the implementation
        ListNode* dummy = new ListNode(0);
        ListNode* current = dummy;
        int carry = 0;

        // Traverse both linked lists
        while (l1 != nullptr || l2 != nullptr || carry != 0) {
            int sum = carry; // Start with the carry value

            if (l1 != nullptr) {
                sum += l1->val; // Add l1's value if it exists
                l1 = l1->next;  // Move to the next node
            }

            if (l2 != nullptr) {
                sum += l2->val; // Add l2's value if it exists
                l2 = l2->next;  // Move to the next node
            }

            carry = sum / 10; // Calculate the new carry
            current->next = new ListNode(sum % 10); // Add the digit to the result
            current = current->next; // Move to the next node
        }

        return dummy->next; // Return the result list, skipping the dummy node
    }
};

File: python/jan/25_24/add_two_numbers.py
class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: Optional[ListNode]
        :type l2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        # Initialize a dummy node to simplify result construction
        dummy = ListNode(0)
        current = dummy
        carry = 0

        # Traverse both linked lists until both are exhausted
        while l1 or l2 or carry:
            # Get values from current nodes, or 0 if the list is exhausted
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            
            # Compute sum and carry
            total = val1 + val2 + carry
            carry = total // 10  # Carry for the next digit
            current.next = ListNode(total % 10)  # Add new digit to result
            
            # Move to the next nodes in l1 and l2
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
            current = current.next

        # Return the next node of the dummy, which is the head of the result
        return dummy.next
